matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -183,18 +183,29 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
         // now find each of the entries in the graph:
         final EdgeFilter edgeFilter = new DefaultEdgeFilter(algoOptions.getWeighting().getFlagEncoder());
 
-        List<List<QueryResult>> queriesPerEntry = lookupGPXEntries(filteredGPXEntries, edgeFilter);
+        List<Collection<QueryResult>> queriesPerEntry =
+                lookupGPXEntries(filteredGPXEntries, edgeFilter);
 
-        // now look up the entries up in the graph:
+        // Add virtual nodes and edges to the graph so that candidates on edges can be represented
+        // by virtual nodes.
         final QueryGraph queryGraph = new QueryGraph(routingGraph).setUseEdgeExplorerCache(true);
         List<QueryResult> allQueryResults = new ArrayList<>();
-        for (List<QueryResult> qrs: queriesPerEntry)
+        for (Collection<QueryResult> qrs: queriesPerEntry)
             allQueryResults.addAll(qrs);
         queryGraph.lookup(allQueryResults);
 
+        // Different QueryResults can have the same tower node as their closest node.
+        // Hence, we now dedupe the query results of each GPX entry by their closest node.
+        // This must be done after calling queryGraph.lookup() since this replaces some of the
+        // QueryResult nodes with virtual nodes. Virtual nodes are not deduped since there is at
+        // most one QueryResult per edge and virtual nodes are inserted into the middle of an edge.
+        // Reducing the number of QueryResults improves performance since less shortest/fastest
+        // routes need to be computed.
```

Comment:
Since this PR does not contain additional information I don't see the benefit of this. But I will do this if you generally prefer linking to PRs from the code.